cli/azd/internal/cmd/add/add_configure.go

```diff
@@ -0,0 +1,209 @@
+package add
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"unicode"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/project"
+	"github.com/fatih/color"
+)
+
+type promptCtx struct {
+	context.Context
+	prj *project.ProjectConfig
+}
+
+// configure fills in the fields for a resource.
+func configure(
+	r *project.ResourceConfig,
+	console input.Console,
+	ctx promptCtx) (*project.ResourceConfig, error) {
+	switch r.Type {
+	case project.ResourceTypeHostContainerApp:
+		return fillUses(r, console, ctx)
+	case project.ResourceTypeOpenAiModel:
+		return fillAiModelName(r, console, ctx)
+	case project.ResourceTypeDbPostgres,
+		project.ResourceTypeDbMongo:
+		return fillDatabaseName(r, console, ctx)
+	case project.ResourceTypeDbRedis:
+		r.Name = "redis"
+		return r, nil
+	default:
+		return r, nil
+	}
+}
+
+func fillDatabaseName(
+	r *project.ResourceConfig,
+	console input.Console,
+	ctx promptCtx) (*project.ResourceConfig, error) {
+	if r.Name != "" {
+		return r, nil
+	}
+
+	for {
+		dbName, err := console.Prompt(ctx, input.ConsoleOptions{
+			Message: fmt.Sprintf("Input the name of the app database (%s)", r.Type.String()),
+			Help: "Hint: App database name\n\n" +
+				"Name of the database that the app connects to. " +
+				"This database will be created after running azd provision or azd up.",
+		})
+		if err != nil {
+			return r, err
+		}
+
+		if err := validateResourceName(dbName, ctx.prj); err != nil {
+			console.Message(ctx, err.Error())
+			continue
+		}
+
+		r.Name = dbName
+		break
+	}
+
+	return r, nil
+}
+
+func fillAiModelName(
+	r *project.ResourceConfig,
+	console input.Console,
+	ctx promptCtx) (*project.ResourceConfig, error) {
+	if r.Name != "" {
+		return r, nil
+	}
+
+	modelProps, ok := r.Props.(project.AIModelProps)
+	defaultName := ""
+
+	// provide a default suggestion using the underlying model name
+	if ok {
+		defaultName = modelProps.Model.Name
+		i := 1
+		for {
+			if _, exists := ctx.prj.Resources[defaultName]; exists {
+				i++
+				defaultName = fmt.Sprintf("%s-%d", defaultName, i)
+			} else {
+				break
+			}
+		}
+	}
+
+	for {
+		modelName, err := console.Prompt(ctx, input.ConsoleOptions{
+			Message:      "What is the name of this model?",
+			DefaultValue: defaultName,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		if err := validateResourceName(modelName, ctx.prj); err != nil {
+			console.Message(ctx, err.Error())
+			continue
+		}
+
+		r.Name = modelName
+		break
+	}
+
+	return r, nil
+}
+
+func fillUses(
+	r *project.ResourceConfig,
+	console input.Console,
+	ctx promptCtx) (*project.ResourceConfig, error) {
+	type resourceDisplay struct {
+		Resource *project.ResourceConfig
+		Display  string
+	}
+	res := make([]resourceDisplay, 0, len(ctx.prj.Resources))
+	for _, r := range ctx.prj.Resources {
```

Comment:
I'm concern about using `ctx.prj`  the way it currently is.  Since it is a pointer, there is a big assumption about the caller will aways make sure to set this pointer, but if that doesn't happen, there will be a `panic` on runtime.

Are you creating a context type instead of using `context.WithValue(parent, key, value)` ??  I think that might be a more idiomatic way to work with just one context type.

Reply:
> which not what we want here.

Why not?  I think what we want is to use a context as a backpack to carry data across the stack (calling one method after another) without making all methods to know about all those parameters.
For example, consider methods A, B, C and D.   A will call B, then B calls C, and finally C calls D.
Now, D requires param `foo`, but neither `C` or `B` uses `foo`.
By using a context backpack, A can call `withValue(ctx, fooKey, foo)` and call B(ctx).
When `D` is finally called, it will try to pull `foo` from the `ctx`.   It will panic or fail if `foo` is not in the ctx.

The composite approach from Wei comes with some type safety, but it would still let `B` and `C` to see and even update/mutate the current context... which I think shouldn't happen.   If B or C would like to change the scope for `D`, they should create a new ctx with a new `foo` :)

Reply:
Separated out as options for now